cmd/clusterctl/hack/local-overrides.py

```diff
@@ -0,0 +1,181 @@
+        try:
+            os.makedirs(provider_overrides_folder)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise
+        f = open(os.path.join(provider_overrides_folder, components_file), 'w')
+        f.write(components_yaml)
+        f.close()
+    except Exception as e:
+        raise Exception('failed to write {} to {}: {}'.format(components_file, components_folder, e))
+
+def create_local_overrides():
+    providerList = settings.get('providers', [])
+    assert providerList is not None, 'invalid configuration: please define the list of providers to override'
+    assert len(providerList)>0, 'invalid configuration: please define at least one provider to override'
+
+    for provider in providerList:
+        p = providers.get(provider)
+        assert p is not None, 'invalid configuration: please specify the configuration for the {} provider'.format(provider)
+
+        repo = p.get('repo', '.')
+        config_folder = p.get('configFolder', '/config/default')
+
+        next_version = p.get('nextVersion')
+        assert next_version is not None, 'invalid configuration for provider {}: please provide nextVersion value'.format(provider)
+
+        type = p.get('type')
+        assert type is not None, 'invalid configuration for provider {}: please provide type value'.format(provider)
+        assert type in validTypes, 'invalid configuration for provider {}: please use one of {}'.format(provider, ', '.join(validTypes))
+
+        components_file = p.get('componentsFile')
+        assert components_file is not None, 'invalid configuration for provider {}: please provide componentsFile value'.format(provider)
+
+        components_yaml = execCmd(['kustomize', 'build', repo + config_folder])
+        write_local_override(provider, next_version, components_file, components_yaml)
+
+        yield provider, type, next_version
+
+
+def CoreProviderFlag():
+    return '--core'
+
+def BootstrapProviderFlag():
+    return '--bootstrap'
+
+def InfrastructureProviderFlag():
+    return '--infrastructure'
+
+def type_to_flag(type):
+    switcher = {
+        'CoreProvider': CoreProviderFlag,
+        'BootstrapProvider': BootstrapProviderFlag,
+        'InfrastructureProvider': InfrastructureProviderFlag
+    }
+    func = switcher.get(type, lambda: 'Invalid type')
+    return func()
+
+def print_instructions(overrides):
+    providerList = settings.get('providers', [])
+    print ('clusterctl local overrides generated from local repositories for the {} providers.'.format(', '.join(providerList)))
+    print ('in order to use them, please run:')
+    print
+    cmd = 'clusterctl init'
+    for provider, type, next_version in overrides:
+        cmd += ' {} {}:{}'.format(type_to_flag(type), provider, next_version)
+    print (cmd)
+    print
+
+load_settings()
+
+load_providers()
+
+overrides = create_local_overrides()
+
+print_instructions(overrides)
```

Comment:
Since you're going to have to squash this before we lgtm it, could you please add a newline at the end of the file? (most editors/IDEs have a setting to do this automatically, if you want)

Reply:
done